ts/session/sending/PendingMessageCache.ts

```diff
@@ -1,36 +1,144 @@
-    // TODO: Return previous pending message if it exists
-    return {} as RawMessage;
+  public get(): Array<RawMessage> {
+    // Get all pending from cache, sorted with oldest first
+    return this.cache.sort((a, b) => a.timestamp - b.timestamp);
   }
 
-  public removePendingMessage(message: RawMessage) {
-    // TODO: implement
+  public async add(
+    device: PubKey,
+    message: ContentMessage
+  ): Promise<RawMessage> {
+    const rawMessage = MessageUtils.toRawMessage(device, message);
+
+    // Does it exist in cache already?
+    if (this.find(rawMessage)) {
+      return rawMessage;
+    }
+
+    this.cache.push(rawMessage);
+    await this.syncCacheWithDB();
+
+    return rawMessage;
+  }
+
+  public async remove(
+    message: RawMessage
+  ): Promise<Array<RawMessage> | undefined> {
+    // Should only be called after message is processed
+
+    // Return if message doesn't exist in cache
+    if (!this.find(message)) {
+      return;
+    }
+
+    // Remove item from cache and sync with database
+    const updatedCache = this.cache.filter(
+      m => m.identifier !== message.identifier
+    );
+    this.cache = updatedCache;
+    await this.syncCacheWithDB();
+
+    return updatedCache;
+  }
+
+  public find(message: RawMessage): RawMessage | undefined {
+    // Find a message in the cache
+    return this.cache.find(
+      m => m.device === message.device && m.timestamp === message.timestamp
+    );
+  }
+
+  public getForDevice(device: PubKey): Array<RawMessage> {
+    const pending = this.cache.filter(m => m.device === device.key);
+
+    return pending.sort((a, b) => a.timestamp - b.timestamp);
   }
 
-  public getPendingDevices(): Array<String> {
-    // TODO: this should return all devices which have pending messages
-    return [];
+  public async clear() {
+    // Clears the cache and syncs to DB
+    this.cache = [];
+    await this.syncCacheWithDB();
+  }
+
+  public getDevices(): Array<PubKey> {
+    // Gets all unique devices with pending messages
+    const pubkeyStrings = [...new Set(this.cache.map(m => m.device))];
+
+    const pubkeys: Array<PubKey> = [];
+    pubkeyStrings.forEach(pubkey => {
+      if (PubKey.validate(pubkey)) {
+        pubkeys.push(new PubKey(pubkey));
+      }
+    });
+
+    return pubkeys;
+  }
+
+  public async init() {
+    const messages = await this.getFromStorage();
+    this.cache = messages;
+  }
+
+  private async getFromStorage(): Promise<Array<RawMessage>> {
+    // tslint:disable-next-line: no-backbone-get-set-outside-model
+    const data = await getItemById('pendingMessages');
+    if (!data || !data.value) {
+      return [];
+    }
+
+    const barePending = JSON.parse(String(data.value));
+
+    // tslint:disable-next-line: no-unnecessary-local-variable
+    const pending = barePending.map((message: any) => {
+      const { identifier, timestamp, device, ttl, encryption } = message;
+
+      // Recreate buffers
+      const rawBuffer = message.plainTextBuffer;
```

Comment:
if this is an `ArrayBuffer` you can just do `new Uint8Array(buffer)`

Reply:
It's just an object of {0: number, 1: number ...}